cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll

```diff
@@ -0,0 +1,48 @@
+/**
+ * Provides implementation classes modeling `select` and various similar
+ * functions. See `semmle.code.cpp.models.Models` for usage information.
+ */
+
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The function `select` and its assorted variants
+ */
+private class Select extends ArrayFunction, AliasFunction, TaintFunction, SideEffectFunction {
+  Select() { this.hasGlobalName(["select", "pselect"]) }
+
+  override predicate hasArrayWithUnknownSize(int bufParam) { bufParam = [1 .. 3] }
```

Comment:
Do you not think `nfds` is a good enough expression of the array size?

Reply:
I don't think so, no. According this [this](https://man7.org/linux/man-pages/man2/select.2.html) description:
> This argument should be set to the highest-numbered file descriptor in any of the three sets, plus 1.

So I guess `fds` could point to a single-element array containing a file descriptor with the value 42, and thus you should set `nfds` to 43. I admit, this is a weird way of doing things, but I'm sure they have/had a good reason for this.

Reply:
I thought `ndfs` was simply the size of the three pointed-to arrays, but you're right - the parameters are in fact pointers to `fd_set`, a struct containing either a fixed bitmap or a resizable array.  The exact purpose of `ndfs` is still unclear to me, but it isn't the array size, and `hasArrayWithUnknownSize` is appropriate.